File: backend/models/project.go
package models

import (
	"github.com/google/uuid"
	"time"
)

type Project struct {
	ID           uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	CompanyID    uuid.UUID  `gorm:"type:uuid;not null;index" json:"company_id"`
	CariID       uuid.UUID  `gorm:"type:uuid;not null;index" json:"cari_id"`
	Name         string     `gorm:"type:varchar(255);not null" json:"name"`
	Description  string     `gorm:"type:text" json:"description"`
	Code         string     `gorm:"type:varchar(50);not null;uniqueIndex:,composite:company_id" json:"code"`
	Status       string     `gorm:"type:varchar(50);not null;default:'planning'" json:"status"` // 'planning', 'in_progress', 'on_hold', 'completed', 'cancelled'
	CategoryID   *uuid.UUID `gorm:"type:uuid" json:"category_id"`
	Category     *ProjectCategory `gorm:"foreignKey:CategoryID;constraint:OnDelete:SET NULL" json:"category,omitempty"`
	StartDate    time.Time  `gorm:"not null" json:"start_date"`
	EndDate      time.Time  `gorm:"not null" json:"end_date"`
	Budget       *float64   `json:"budget"`
	Note         string     `gorm:"type:text" json:"note"`
	Invoices     []Invoice  `gorm:"many2many:project_invoices;constraint:OnDelete:CASCADE" json:"invoices,omitempty"`
	Quotes       []Quote    `gorm:"many2many:project_quotes;constraint:OnDelete:CASCADE" json:"quotes,omitempty"`
	Employees    []Employee `gorm:"many2many:project_employees;constraint:OnDelete:CASCADE" json:"employees,omitempty"`
	CreatedBy    *uuid.UUID `gorm:"type:uuid" json:"created_by"`
	UpdatedBy    *uuid.UUID `gorm:"type:uuid" json:"updated_by"`
	CreatedByUser *User    `gorm:"foreignKey:CreatedBy;constraint:OnDelete:SET NULL" json:"created_by_user,omitempty"`
	UpdatedByUser *User    `gorm:"foreignKey:UpdatedBy;constraint:OnDelete:SET NULL" json:"updated_by_user,omitempty"`
	CreatedAt    time.Time  `json:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at"`
}

File: frontend/src/stores/project.ts
import { defineStore } from 'pinia'
import {
  getProjectsApi,
  getProjectByIDApi,
  createProjectApi,
  updateProjectApi,
  deleteProjectApi,
  addInvoiceToProjectApi,
  removeInvoiceFromProjectApi,
  addQuoteToProjectApi,
  removeQuoteFromProjectApi,
  addEmployeeToProjectApi,
  removeEmployeeFromProjectApi,
} from '@/api/project'

import type { UserRef } from './invoice'

export interface Project {
  id: string
  company_id: string
  cari_id: string
  name: string
  description?: string
  code: string
  category_id?: string | null
  status: 'planning' | 'in_progress' | 'on_hold' | 'completed' | 'cancelled'
  start_date: string
  end_date: string
  budget?: number
  note?: string
  employees?: any[]
  invoices?: any[]
  quotes?: any[]
  category?: any
  created_at: string
  updated_at: string
  created_by_user?: UserRef | null
  updated_by_user?: UserRef | null
}

export const useProjectStore = defineStore('project', {
  state: () => ({
    projects: [] as Project[],
    total: 0,
    activeProject: null as Project | null,
    loading: false,
  }),
  actions: {
    async fetchProjects(params?: any) {
      this.loading = true
      try {
        const res = await getProjectsApi(params)
        this.projects = res.data.data
        this.total = res.data.meta.total
      } finally {
        this.loading = false
      }
    },
    async fetchProjectByID(id: string) {
      this.loading = true
      try {
        const res = await getProjectByIDApi(id)
        this.activeProject = res.data.data
      } finally {
        this.loading = false
      }
    },
    async createProject(data: any) {
      const res = await createProjectApi(data)
      return res.data.data
    },
    async updateProject(id: string, data: any) {
      const res = await updateProjectApi(id, data)
      return res.data.data
    },
    async deleteProject(id: string) {
      await deleteProjectApi(id)
    },
    async addInvoiceToProject(projectId: string, invoiceId: string) {
      const res = await addInvoiceToProjectApi(projectId, invoiceId)
      return res.data.data
    },
    async removeInvoiceFromProject(projectId: string, invoiceId: string) {
      const res = await removeInvoiceFromProjectApi(projectId, invoiceId)
      return res.data.data
    },
    async addQuoteToProject(projectId: string, quoteId: string) {
      const res = await addQuoteToProjectApi(projectId, quoteId)
      return res.data.data
    },
    async removeQuoteFromProject(projectId: string, quoteId: string) {
      const res = await removeQuoteFromProjectApi(projectId, quoteId)
      return res.data.data
    },
    async addEmployeeToProject(projectId: string, employeeId: string) {
      const res = await addEmployeeToProjectApi(projectId, employeeId)
      return res.data.data
    },
    async removeEmployeeFromProject(projectId: string, employeeId: string) {
      const res = await removeEmployeeFromProjectApi(projectId, employeeId)
      return res.data.data
    },
  },
})
